components/core/src/clp/streaming_compression/lzma/Compressor.cpp

```diff
@@ -0,0 +1,214 @@
+    m_compression_stream.avail_out = m_compressed_stream_block_buffer.size();
+
+    m_compressed_stream_file_writer = &file_writer;
+
+    m_uncompressed_stream_pos = 0;
+}
+
+auto Compressor::close() -> void {
+    if (nullptr == m_compressed_stream_file_writer) {
+        throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
+    }
+
+    flush_and_close_compression_stream();
+    m_compressed_stream_file_writer = nullptr;
+}
+
+auto Compressor::write(char const* data, size_t data_length) -> void {
+    if (nullptr == m_compressed_stream_file_writer) {
+        throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
+    }
+
+    if (0 == data_length) {
+        // Nothing needs to be done because we do not need to compress anything
+        return;
+    }
+
+    if (nullptr == data) {
+        throw OperationFailed(ErrorCode_BadParam, __FILENAME__, __LINE__);
+    }
+
+    m_compression_stream.next_in = size_checked_pointer_cast<uint8_t const>(data);
+    m_compression_stream.avail_in = data_length;
+
+    // Normal compression encoding workflow. Continue until the input buffer is
+    // exhausted.
+    compress(LZMA_RUN);
+
+    m_compression_stream.next_in = nullptr;
+
+    m_compression_stream_contains_data = true;
+    m_uncompressed_stream_pos += data_length;
+}
+
+auto Compressor::flush() -> void {
+    if (false == m_compression_stream_contains_data) {
+        return;
+    }
+
+    // Forces all the buffered data to be available at output
+    compress(LZMA_SYNC_FLUSH);
+    m_compression_stream_contains_data = false;
+}
+
+auto Compressor::try_get_pos(size_t& pos) const -> ErrorCode {
+    if (nullptr == m_compressed_stream_file_writer) {
+        return ErrorCode_NotInit;
+    }
+
+    pos = m_uncompressed_stream_pos;
+    return ErrorCode_Success;
+}
+
+auto Compressor::flush_and_close_compression_stream() -> void {
+    if (nullptr == m_compressed_stream_file_writer) {
+        throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
+    }
+
+    // Same as flush but all the input data must have been given to the encoder
+    compress(LZMA_FINISH);
+
+    m_compression_stream_contains_data = false;
+
+    lzma_end(&m_compression_stream);
+    m_compression_stream.avail_out = 0;
+    m_compression_stream.next_out = nullptr;
+}
+
+auto Compressor::compress(LzmaAction action) -> void {
+    bool hit_stream_end{false};
+    while (true) {
+        auto const rc = lzma_code(&m_compression_stream, action);
+        switch (rc) {
+            case LZMA_OK:
+            case LZMA_BUF_ERROR:
+                break;
+            case LZMA_STREAM_END:
+                hit_stream_end = true;
+                break;
+            default:
+                SPDLOG_ERROR("lzma() returned an unexpected value - {}.", static_cast<int>(rc));
+                throw OperationFailed(ErrorCode_Failure, __FILENAME__, __LINE__);
+        }
+
+        if (LZMA_RUN == action && 0 == m_compression_stream.avail_in) {
```

Comment:
quick question: what's the difference between
0 == m_compression_stream.avail_in and rc == LZMA_STREAM_END?

Reply:
I've updated the logic a bit.
`0 == m_compression_stream.avail_in` means the input data is exhausted. All input data have been processed to be encoded.
`rc == LZMA_STREAM_END` means flushing is complete. All buffered encoded input data has been flushed to the output

Reply:
Does it mean that rc == LZMA_STREAM_END will only occur if the action is LZMA_SYNC_FLUSH or Finish?

Reply:
In the doc: 
https://github.com/tukaani-project/xz/blob/fe9e66993fdbcc2981c7361b9b034a451eb0fc42/src/liblzma/api/lzma/base.h#L61
It says this marks the finish of `LZMA_SYNC_FLUSH`, `LZMA_FULL_FLUSH`, or `LZMA_FINISH`. If it can occur for `LZMA_RUN` I think they would include it.
The `lzma_code` is a rather complicated state machine.